### ISSUE TITLE
Bulk inserts

### ISSUE THREAD

**Issue by sroddy** [tekartik/sqflite#19]
Currently it's not possible to bulk insert records using a simple API.

I managed to achieve this result by manually creating this kind of query and binding the parameters to it. The gain in speed is remarkable (10x at least):
```
INSERT INTO table (col1, col2) VALUES (?, ?), (?, ?), (?, ?), ...
```
It shouldn't be that hard to implement a bulk insert method that follows this pattern. The only thing to pay attention to is that SQLite limits the max bound arguments to 500 per query, so if the records/columns are a lot, queries need to be splitted and batched (maybe in a transaction)

**Comment by alextekartik** [comment]
I was not aware of this possibility, sorry. You mean 10x over the same inserts in a transaction? Indeed that might be worth the effort. I guess the API should be similar to what we have now for insert and rawInsert. Would having bulkInsert and rawBulkInsert returning a list of ids would be something that matches your need. Otherwise I'm opened for suggestion

**Comment by sroddy** [comment]
I like your approach for the API, even though I don't really need the IDs for my specific use-case as I'm performing an upsert including server-provided IDs too.

Performance wise, my guess is that the bottleneck is caused by the async ping-pong between native and dart, and it's way more noticeable on Android (probably because there is an extra overhead due to JNI?).

It would be probably not that noticeable (or even negligible) if it were performing one query per time while staying on the native side, so before wasting time on a complex implementation like the one I suggested in the first step, could you have a try to implement a method like you suggest that performs a bulk of single inserts in a transaction while staying on the native side?

Thank you very much, I really appreciate your work!

**Comment by alextekartik** [comment]
So I took a different approach and add a global `Batch` support for insert/update/delete, which is similar to the bulkInsert light version you proposed. In a batch you can get the result of each operation, which has cost (especially on Android) so I added a option in the batch.commit to not fetch those result. On Android emulator I can see a x7 improvement compare to a regular transaction which might be explained by the ping-pong you mentioned. Not getting the result itself has a x3 improvement on Android. I don't have much data to test with so if you can tell me if you get any improvement, that would be great.

      batch = db.batch();
      batch.insert("Test", {"name": "item"});
      batch.update("Test", {"name": "new_item"},
          where: "name = ?", whereArgs: ["item"]);
      batch.delete("Test", where: "name = ?", whereArgs: ["item"]);
      await batch.commit(noResult: true);

**Comment by alextekartik** [comment]
Published on pub (v0.4.0)

**Comment by alextekartik** [comment]
Assuming the batch solution is a satisfactory solution for doing some kind of bulk insert.
